### PR TITLE
fix(coverage): treat startswith edges as covering printf-format markers (#141)

### LINKED ISSUE #141: dippin coverage: false-positive 'uncovered output' on printf-format markers consumed via startswith edges
## Summary

`dippin coverage` (the `tool node(s) with uncovered outputs` check that drives `dippin doctor` from A → B) does not reason about `startswith` edge predicates against printf-format strings. It compares the literal-output set of a tool body to the literal-edge set, and flags any tool body literal that no edge exact-matches — even when the edge uses `startswith` against the format-string's static prefix.

This is a sibling of (closed) #40 and (closed) #50, which addressed file-redirected emissions. This issue is about a distinct case: stdout-emitted printf-format markers (e.g. `printf 'citations-fail-%s' "$ids"`) consumed by `... when ctx.tool_stdout startswith citations-fail` edges.

## Concrete repro (live in 2389-research/pipelines)

File: `iterative/iter_scope.dip` after PR https://github.com/2389-research/pipelines/pull/64 lands.

**Case 1 — \`check_citations\`** at `iter_scope.dip:240-263`:
- printf bodies emit:
  - \`citations-ok\` (literal)
  - \`citations-fail-%s\` (format, suffix populated at runtime from \`\$ids\`)
- edges at `iter_scope.dip:579-580`:
  - \`check_citations -> par_scope_dispatch when ctx.tool_stdout = citations-ok\`
  - \`check_citations -> fix_citations when ctx.tool_stdout startswith citations-fail\`
- `dippin coverage` reports: missing \`citations-fail-%s\`.

**Case 2 — \`validate_roadmap\`** at `iter_scope.dip:511-539`:
- printf bodies emit 5 distinct \`roadmap-invalid-<literal-suffix>\` strings (file-missing, missing-walking-skeleton-section, no-iter-sections, missing-status-fields, no-story-refs) plus \`roadmap-valid\`
- edges at `iter_scope.dip:599-600`:
  - \`validate_roadmap -> commit_roadmap when ctx.tool_stdout = roadmap-valid\`
  - \`validate_roadmap -> write_roadmap when ctx.tool_stdout startswith roadmap-invalid\`
- `dippin coverage` reports 5 missing \`roadmap-invalid-*\` literals.

Both cases are routed exhaustively by the \`startswith\` edges; the coverage analyzer doesn't see them as covered because it does literal exact-match instead of evaluating the predicate.

## Impact

- `dippin doctor` ceiling stays at B/80 on otherwise-clean files (DIP101/DIP102 = 0 after PR #64 lands; only the coverage false-positives remain).
- Same false-positive recurs on every `.dip` using the post-#39 `marker_grep:` + \`startswith\` cascade pattern — including `iter_extract.dip` and the spec_to_sprints / iter_dev cluster from pipelines commit \`40ca60c\`.
- Forces \"honest scope note\" boilerplate in PR bodies because the grade gap can't actually be closed.

## Suggested approach

For each printf-format literal in a tool body, check whether *some* outgoing edge has a \`startswith\` (or regex / matches) predicate whose constant prefix is a prefix of the format string's literal portion. If yes, treat it as covered.

A more general (but more invasive) approach: feed the coverage analyzer the same \`marker_grep\` regex the linter already uses — if the regex matches the literal-prefix-of-the-format-string, treat as covered.

## Workaround in the meantime

PR #64's \"Honest scope note\" pattern: ship the lint-clean fix, document the B/80 ceiling, link this issue.

## Related

- Closed siblings: #40 (file-redirected), #50, #39 (marker_grep parser)
- Open umbrella: #127 (edge / routing syntax evolution tracking)

### PR DESCRIPTION
**Pull request by clintecker**
Fixes #141.

## Problem

`dippin coverage` (which drives the `dippin doctor` "tool node(s) with uncovered outputs" check) did exact-match of each tool-output literal against a flat set of edge-condition *values*, discarding the operator. A `... when ctx.tool_stdout startswith citations-fail` edge contributed only the bare literal `citations-fail`, which never exact-matched the printf-format output `citations-fail-%s`. The result was a false-positive "uncovered output" that capped doctor grades on otherwise-clean files using the `startswith`-cascade routing pattern.

## Fix

Carry the operator through. `appendConditionTerms`/`collectEdgeConditionTerms` collect each comparison as a typed `condTerm{op, value}`. `findMissingEdges` now does two-tier matching: an output is covered if some `=`/`==` term equals it (existing behavior) **or** some `startswith` term is a prefix of it (new). Conservative by design — only `startswith` is extended; `contains`/`endswith`/regex/`marker_grep` are left out to favor false negatives over crying wolf.

The public `EdgeConditions []string` report field is preserved unchanged (still the flat value list, asserted by the existing `TestCollectEdgeConditions_*` tests).

## Tests

- `examples/coverage_startswith.dip` — regression fixture (passes `dippin validate`; covered by the validate-examples pre-commit gate and `TestLintExamples`).
- `TestCoverageStartswith_FixtureCovered` — parses the fixture, asserts `citations-fail-%s` is no longer in `MissingEdges` and the node status is `covered`.
- `TestFindMissingEdges_Startswith` / `_StartswithNonPrefix` — unit coverage of the matcher (prefix match covers; non-prefix stays missing).

`scripts/pre-commit` passes end-to-end (build, vet, golangci-lint, gofmt, full test suite, release checks, cyclomatic/cognitive complexity, validate-examples).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/2389-research/codesmith/dippin-lang/pr/147"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1783888082&installation_id=131176874&pr_number=147&repository=2389-research%2Fdippin-lang&return_to=https%3A%2F%2Fgithub.com%2F2389-research%2Fdippin-lang%2Fpull%2F147&signature=bf000fb7bc9d4a1d9ddb0482f7d492029f988a01eef4896ad06c388db699bcfe"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: coverage/coverage_test.go [hunk 1]
```diff
@@ -598,3 +600,56 @@ func slicesEqual(a, b []string) bool {
...
+	missing := findMissingEdges(outputs, terms)
+	if len(missing) != 0 {
+		t.Errorf("expected no missing edges, got %v", missing)
+	}
+}
+
+// TestFindMissingEdges_StartswithNonPrefix asserts conservatism: a
+// `startswith` term that is NOT a prefix of the output leaves it missing.
+func TestFindMissingEdges_StartswithNonPrefix(t *testing.T) {
+	outputs := []string{"citations-fail-%s"}
```
Comment (clintecker) [inline coverage/coverage_test.go:624]: **P2 (missing test):** The new `appendConditionTerms` recursion into `CondAnd`/`CondOr`/`CondNot` is the part that has to preserve the operator through the tree, but every new test here uses a bare top-level `CondCompare` (and the fixture's `startswith` edge is also a flat compare). The realistic routing pattern this PR targets is compound — e.g. `... when ctx.outcome = fail and ctx.tool_stdout startswith err-`. I verified the recursion does work for that case, but nothing in the suite guards it: a future refactor that, say, only extracts the `Left` branch or collapses operators would pass all tests. Suggest adding one test asserting `outputCovered("err-database", appendConditionTerms(nil, CondAnd{...startswith...}))` is true. The existing `TestAppendConditionValues_CondAnd/Or/Not` cover the *old* flat path, not this new one.

File: coverage/coverage.go [hunk 2]
```diff
@@ -324,6 +347,40 @@ func collectEdgeConditions(edges []*ir.Edge) (conditions []string, hasFallback b
...
+	switch e := expr.(type) {
+	case ir.CondCompare:
+		return append(out, condTerm{op: e.Op, value: e.Value})
+	case ir.CondAnd:
+		out = appendConditionTerms(out, e.Left)
+		return appendConditionTerms(out, e.Right)
+	case ir.CondOr:
+		out = appendConditionTerms(out, e.Left)
+		return appendConditionTerms(out, e.Right)
+	case ir.CondNot:
```
Comment (clintecker) [inline coverage/coverage.go:377]: **P3 (informational, pre-existing):** `appendConditionTerms` recurses into `CondNot` and emits the inner comparison as a *positive* coverage term, discarding the negation. So `... when not ctx.tool_stdout startswith err-` would now mark every `err-`-prefixed output as covered, even though that edge fires for the *complement*. This flattening is inherited from `appendConditionValues` (and is explicitly asserted by `TestCollectEdgeConditions_NotCondition`), so it's not introduced by this PR and negated-startswith routing is rare — flagging only so it's a known limitation, not a merge blocker.

File: coverage/coverage.go [hunk 3]
```diff
@@ -324,6 +347,40 @@ func collectEdgeConditions(edges []*ir.Edge) (conditions []string, hasFallback b
...
+// value) from edges, preserving the operator so prefix predicates can be
+// matched. The flat EdgeConditions report field is produced separately by
+// collectEdgeConditions.
+func collectEdgeConditionTerms(edges []*ir.Edge) []condTerm {
+	var terms []condTerm
+	for _, e := range edges {
+		if e.Condition == nil {
+			continue
+		}
+		terms = appendConditionTerms(terms, e.Condition.Parsed)
```
Comment (clintecker) [inline coverage/coverage.go:360]: **P3 (minor cleanup/efficiency):** `analyzeToolNode` now walks each node's edges twice — once via `collectEdgeConditions` (for the flat `EdgeConditions` report field + `hasFallback`) and once via `collectEdgeConditionTerms` (for matching). The two could share a single edge pass that returns both the flat values and the typed terms. Negligible cost in practice; only worth doing if you touch this again. Not a blocker.

**Review by clintecker — COMMENTED** [review]
## Review: fix(coverage): startswith edges cover printf-format markers (#141)

Verified locally on a detached scratch worktree at the PR head (the branch is already checked out in a locked agent worktree, which I left untouched). `scripts/pre-commit` passes end-to-end: build, vet, golangci-lint (0 issues), gofmt, full test suite, release checks, **cyclomatic + cognitive complexity gates**, and validate-examples. No expected `tree-sitter-generate` failure surfaced this run.

**The fix is correct and addresses the issue.** I independently confirmed:
- `termCovers` matches in the right direction (`HasPrefix(output_literal, edge_constant)`), so `citations-fail-%s` is covered by `startswith citations-fail`.
- The node `Status` flips `partial` -> `covered`, which removes the 10-point/node penalty `doctor.countUncoveredTools` applies (doctor.go:111,127) — exactly the B/80 ceiling described in #141.
- The public `EdgeConditions []string` report field is **preserved unchanged** (still produced by the separate `collectEdgeConditions`; existing `TestCollectEdgeConditions_*` still green).
- The new fixture is **format-idempotent** and **round-trips** (parse->format->re-parse->validate), so it passes the CI `validate-examples` idempotency + round-trip gates.
- Conservatively handling ONLY `startswith` (and dropping `!=` from positive coverage vs. the old flat-value path, which is strictly more correct) is the right call — favoring false-negatives over crying wolf.

I checked and **refuted** two candidate defects before posting: (a) an empty `startswith` constant (`HasPrefix(x,"")` always true → would mask everything) is unreachable because the parser rejects an empty value with DIP010; (b) the inverted-prefix direction is correct.

Findings below are advisory (a missing-test gap + two informational notes); none block merge.

### P2
- Missing unit test for compound conditions carrying a `startswith` term through the new `appendConditionTerms` recursion.

### P3
- `CondNot` inverts predicate meaning but is flattened as positive coverage (pre-existing, now also affects `startswith`).
- Edges are walked twice per node.
